Frameworks/Foundation/NSUndoManager.mm

```diff
@@ -38,67 +102,159 @@ - (id)prepareWithInvocationTarget:(id)target {
 }
 
 /**
- @Status Stub
- @Notes
+ @Status Interoperable
 */
 - (void)undo {
-    UNIMPLEMENTED();
+    _groupingLevel = [_undoStack count];
+    if (_groupingLevel == 1) {
+        [self endUndoGrouping];
+    }
+
+    [self undoNestedGroup];
 }
 
 /**
- @Status Stub
- @Notes
+ @Status Interoperable
 */
 - (void)undoNestedGroup {
-    UNIMPLEMENTED();
+    _groupingLevel = [_undoStack count];
+
+    if (![self canUndo]) {
+        return;
+    }
+
+    if (_groupingLevel == 0) {
+        [NSException raise:NSInternalInconsistencyException format:@"No top level group to undo."];
+    }
+
+    _undoing = YES;
+
+    _NSUndoGroup* undoGroup = [_undoStack peek];
+
+    // Record redo operation as a single redo group
+    [self _beginRedoGrouping];
+    [undoGroup undo];
+    [self _endRedoGrouping];
+    [_undoStack pop];
+
+    _groupingLevel = [_undoStack count];
+    ;
+
+    _undoing = NO;
 }
 
 /**
- @Status Stub
- @Notes
+ @Status Interoperable
 */
 - (void)redo {
-    UNIMPLEMENTED();
+    _groupingLevel = [_redoStack count];
+    if (![self canRedo]) {
+        return;
+    }
```

Comment:
Because you don't want to adjust the grouping level if you're about to bail out.

---
In reply to: [76160720](https://github.com/Microsoft/WinObjC/pull/900#discussion_r76160720) [](ancestors = 76160720,76160709)